airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLFormatFileCreator.kt

```diff
@@ -0,0 +1,232 @@
+            val match =
+                dbColumns.firstOrNull { it.name.equals(csvColName, ignoreCase = true) }
+                    ?: error(
+                        "No matching DB column found for CSV column '$csvColName' in table schema.",
+                    )
+            CsvToDbColumn(
+                csvPosition = idx + 1,
+                dbOrdinal = match.ordinalPosition,
+                dbColumnName = match.name,
+                dbDataType = match.dataType,
+                dbCharLength = match.charMaxLength,
+            )
+        }
+    }
+
+    /**
+     * Builds the content of the non-XML format file (.fmt) for SQL Server’s BULK INSERT. This ties
+     * CSV columns (SQLCHAR, etc.) to table columns in the correct order, using UTF-8 collation and
+     * automatically determining a prefix length if the column is nullable.
+     */
+    internal fun buildFormatFileContent(
+        csvToDbMapping: List<CsvToDbColumn>,
+        delimiter: String,
+        rowDelimiter: String,
+        formatFileVersion: String
+    ): String {
+        val sb = StringBuilder()
+
+        // First line: format file version (e.g. 12.0)
+        sb.appendLine(formatFileVersion)
+        // Second line: total field count
+        sb.appendLine(csvToDbMapping.size.toString())
+
+        // Next N lines: define each column.
+        // Format:
+        // <CSVFieldPos> <BCPType> <PrefixLength> <Length> <Delimiter> <ServerColumnOrder>
+        // <ColumnName> <Collation>
+        csvToDbMapping.forEachIndexed { i, mapping ->
+            val (csvPos, dbOrdinal, dbColName, dbDataType, dbCharLength) = mapping
+
+            val (bcpType, length) = pickBcpTypeAndLength(dbDataType, dbCharLength)
+            val isLast = (i == csvToDbMapping.size - 1)
+            val usedDelimiter = if (isLast) "\"$rowDelimiter\"" else "\"$delimiter\""
+
+            // For simplicity, apply the UTF-8 collation to all text-based columns (SQLCHAR).
+            // If the column is numeric or date, the collation will be ignored by the server anyway.
+            sb.appendLine(
+                "%-8d %-10s %-8d %-8d %-8s %-8d %s %s".format(
+                    csvPos,
+                    bcpType,
+                    0,
+                    length,
+                    usedDelimiter,
+                    dbOrdinal,
+                    dbColName,
+                    COLLATION,
+                ),
+            )
+        }
+
+        return sb.toString()
+    }
+
+    /**
+     * Chooses the correct BCP type ("SQLCHAR") and a reasonable max length in bytes based on the
+     * SQL data type. Adapt or expand this logic as needed.
+     */
+    private fun pickBcpTypeAndLength(dataType: String, charMaxLength: Int?): Pair<String, Int> {
+        // For CSV-based imports, it's common to treat all columns as text (SQLCHAR).
+        // The length is a best-guess upper bound in bytes.
+        return when (dataType.lowercase()) {
```

Comment:
Any chance we could align this better on Mssql types?
My concern is that implementations may drift and it could be easier for us to cause unnecessary breaking changes between the two insert modes. My intuition is that a user that started on sql inserts (simple setup) and moves to bulk (for performance) should not have to deal with breaking changes.

Reply:
I think since the table creation logic is the same, the final data type of the table would be the same for both the loads. This is specific mapping of sql data type to format file data type. I dont think moving from one sync mode to another would cause this to break.